xmtp_mls/src/groups/message_history.rs

```diff
@@ -129,37 +198,62 @@ where
             log::error!("error publishing sync group intents: {:?}", err);
         }
 
-        Ok(pin_code)
+        Ok((request_id, pin_code))
     }
 
     pub(crate) async fn send_history_reply(
         &self,
         contents: MessageHistoryReply,
-    ) -> Result<(), GroupError> {
+    ) -> Result<(), MessageHistoryError> {
         // find the sync group
         let conn = self.store().conn()?;
-        let sync_group_id = conn
-            .find_sync_groups()?
-            .pop()
-            .ok_or(GroupError::GroupNotFound)?
-            .id;
-        let sync_group = self.group(sync_group_id)?;
+        let (_, sync_group) = self.get_sync_group()?;
+
+        // sync the group
+        Box::pin(sync_group.sync(self)).await?;
+
+        let messages = sync_group.find_messages(
+            Some(GroupMessageKind::Application),
+            None,
+            None,
+            None,
+            None,
+        )?;
+
+        let last_message = match messages.last() {
+            Some(msg) => {
+                let message_history_content =
+                    serde_json::from_slice::<MessageHistoryContent>(&msg.decrypted_message_bytes);
+                match message_history_content {
+                    Ok(MessageHistoryContent::Request(request)) => {
+                        // check that the request ID matches
+                        if !request.request_id.eq(&contents.request_id) {
+                            return Err(MessageHistoryError::ReplyRequestIdMismatch);
+                        }
+                        Some(msg)
+                    }
+                    Ok(MessageHistoryContent::Reply(_)) => {
+                        // if last message is a reply, it's already been processed
+                        return Err(MessageHistoryError::ReplyAlreadyProcessed);
+                    }
+                    _ => None,
+                }
+            }
+            None => {
+                return Err(MessageHistoryError::NoPendingRequest);
+            }
+        };
+
+        if let Some(msg) = last_message {
+            // ensure the requester is a member of all the groups
+            self.ensure_member_of_all_groups(msg.sender_inbox_id.clone())
+                .await?;
+        }
 
         // the reply message
-        let content_bytes = format!(
-            "{}{DELIMITER}{:?}{DELIMITER}{:?}{DELIMITER}{:?}",
-            contents.url,
-            contents
-                .encryption_key
-                .as_ref()
-                .ok_or(MessageProcessingError::InvalidPayload)?,
-            contents
-                .signing_key
-                .as_ref()
-                .ok_or(MessageProcessingError::InvalidPayload)?,
-            contents.bundle_hash.as_slice()
-        )
-        .into_bytes();
+        let content = MessageHistoryContent::Reply(contents.clone());
+        let content_bytes = serde_json::to_vec(&content)
+            .map_err(|e| MessageHistoryError::Generic(format!("{e}")))?;
```

Comment:
instead of `map_err(/* ... */) and a Generic history message, you can add a variant to MessageHistoryError:


```
#[derive(Debug, Error)]
enum MessageHistoryError {
    #[error("Serialization/Deserialization Error {0}")]
    Serde(#[from] serde_json::Error)
}
```

then you can take advantage of an implicit `From` implementation on the `?` operator to clean up this line a bit:

```rust
let content_bytes = serde_json::to_vec(&content)?;
```


If an error occurs, this will auto-convert the error to `MessageHistory::Serde` variant